packages/unified-selection/src/unified-selection/HiliteSetCache.ts

```diff
@@ -0,0 +1,98 @@
+ *--------------------------------------------------------------------------------------------*/
+
+/** @packageDocumentation
+ * @module UnifiedSelection
+ */
+
+import { from, Observable, shareReplay } from "rxjs";
+import { eachValueFrom } from "rxjs-for-await";
+import { HiliteSet, HiliteSetProvider } from "./HiliteSetProvider";
+import { IMetadataProvider } from "./queries/ECMetadata";
+import { IECSqlQueryExecutor } from "./queries/ECSqlCore";
+import { StorageSelectionChangeEventArgs } from "./SelectionChangeEvent";
+import { SelectionStorage } from "./SelectionStorage";
+
+/**
+ * Caches hilite set for the current `SelectionStorage` selection
+ * so subsequent requests for the same input does not cost
+ * @beta
+ */
+export interface HiliteSetCache {
+  /** Get the current hilite set iterator for the specified imodel */
+  getHiliteSet(props: {
+    /** iModel to get hilite set for */
+    iModelKey: string;
+    /** ECSql query executor */
+    queryExecutor: IECSqlQueryExecutor;
+    /** EC metadata provider */
+    metadataProvider: IMetadataProvider;
+  }): AsyncIterableIterator<HiliteSet>;
+
+  /** Clears hilite set cache for a given iModel */
+  clearCache(props: {
+    /** Key of the iModel to change selection for */
+    iModelKey: string;
+  }): void;
+
+  /** Disposes the cache. */
+  dispose(): void;
+}
+
+/**
+ * Creates a hilite set cache. When an iModel is closed `HiliteSetCache.clearCache` function should be called.
+ * @beta
+ */
+export function createCache(storage: SelectionStorage): HiliteSetCache {
+  return new HiliteSetCacheImpl(storage);
+}
+
+class HiliteSetCacheImpl implements HiliteSetCache {
+  private _selectionStorage: SelectionStorage;
+  private _hiliteSetProviders = new Map<string, HiliteSetProvider>();
+  private _cache = new Map<string, Observable<HiliteSet>>();
+  private _removeListener: () => void;
+
+  constructor(selectionStorage: SelectionStorage) {
+    this._selectionStorage = selectionStorage;
+    this._removeListener = this._selectionStorage.selectionChangeEvent.addListener((args: StorageSelectionChangeEventArgs) =>
+      this._cache.delete(args.iModelKey),
+    );
+  }
+
+  public getHiliteSet(props: { iModelKey: string; queryExecutor: IECSqlQueryExecutor; metadataProvider: IMetadataProvider }): AsyncIterableIterator<HiliteSet> {
+    const { iModelKey, queryExecutor, metadataProvider } = props;
+    const provider = this.getHiliteSetProvider(iModelKey, queryExecutor, metadataProvider);
+    let hiliteSet = this._cache.get(iModelKey);
+
+    if (!hiliteSet) {
+      const selection = this._selectionStorage.getSelection({ iModelKey });
+      hiliteSet = from(provider.getHiliteSet(selection)).pipe(shareReplay({ refCount: true }));
+      this._cache.set(iModelKey, hiliteSet);
+    }
+
+    return eachValueFrom(hiliteSet);
+  }
+
+  public clearCache({ iModelKey }: { iModelKey: string }): void {
+    this._hiliteSetProviders.delete(iModelKey);
+    this._cache.delete(iModelKey);
+  }
```

Comment:
I don't think we need a `clearCache` function - on iModel close, whoever created the `SelectionStorage`, should call `clearStorage`, which will cause a selection clear for that iModel, at which point we're going to get a selection change event and clear the cache.

Reply:
The event would cause the hilite set cache to be cleared, however it would not clear the `_hiliteSetProviders` map, which is why I added this function.

Reply:
Do we need to cache providers?

Reply:
They cache mappings between class names and the types they resolve to that are needed for the queries. Other than that - no.

Reply:
I think those class / type mappings justify caching providers. But I would still like to avoid requiring a `clearCache` call from our consumers... What do you think about using the `StorageSelectionChangeEventArgs.source` attribute to find out whether it's a regular "clear" event or a "clear because of iModel close" event?

Reply:
That was an option I considered at first but decided against it, because the source attribute comes from what is provided to the public `SelectionStorage` methods so it seemed like a hacky and fragile solution.

Reply:
The string comes from the same package where we handle it, so we can use some internal variable to make sure changing it in one place doesn't break the other. If we make it enough unique, we shouldn't have issues with the same string being used from elsewhere. Example: `Unified selection storage: clear`.

Reply:
Removed `clearCache` and switched to clearing based on event source.